org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/BundleUtils.java

```diff
@@ -37,112 +37,141 @@
+
+  private static class BundleInfo {
+
+    private String version;
+
+    private String symbolicName;
+
+    private BundleInfo(String bundleVersion, String symbolicName) {
+      this.version = bundleVersion;
+      this.symbolicName = symbolicName;
+    }
+
+    private String getVersion() {
+      return version;
+    }
+
+    private String getSymbolicName() {
+      return symbolicName;
+    }
+  }
+
+  /**
+   * Load a collection of bundle based on the provided file path locations.
+   *
+   * @param bundleLocations
+   *            The collection of the bundle file path location
+   * @throws CoreException
+   *             throw the <code>CoreException</code> if failed to load any
+   *             bundles.
+   */
+  public static void loadBundles(Collection<String> bundleLocations) throws CoreException {
+    if (bundleLocations == null || bundleLocations.isEmpty()) {
+      return;
+    }
+
+    BundleContext context = JavaLanguageServerPlugin.getBundleContext();
+    MultiStatus status = new MultiStatus(context.getBundle().getSymbolicName(), IStatus.OK, "Load bundle list", null);
+    Collection<Bundle> bundlesToStart = new ArrayList<>();
+    for (String bundleLocation : bundleLocations) {
+      try {
+        if (StringUtils.isEmpty(bundleLocation)) {
+          JavaLanguageServerPlugin.logError("Empty bundle location");
+          continue;
+        }
+
+        String location = getBundleLocation(bundleLocation, true);
+
+        BundleInfo bundleInfo = getBundleInfo(bundleLocation);
+        if (bundleInfo == null) {
+          status.add(new Status(IStatus.ERROR, context.getBundle().getSymbolicName(), "Failed to get bundleInfo for bundle from " + bundleLocation, null));
+          continue;
+        }
+        Bundle bundle = Platform.getBundle(bundleInfo.getSymbolicName());
+        if (bundle != null) {
+          if (bundle.getLocation().equals(location) && bundle.getVersion().equals(Version.parseVersion(bundleInfo.getVersion()))) {
+            // Same bundle exists
+            continue;
+          }
+          bundle.uninstall();
+        }
+        bundle = context.installBundle(location);
+        JavaLanguageServerPlugin.logInfo("Installed " + bundle.getLocation());
+        bundlesToStart.add(bundle);
+
+      } catch (BundleException e) {
+        status.add(new Status(IStatus.ERROR, context.getBundle().getSymbolicName(), "Install bundle failure " + bundleLocation, e));
+      } catch (MalformedURLException ex) {
+        status.add(new Status(IStatus.ERROR, context.getBundle().getSymbolicName(), "Bundle location format is not correct " + bundleLocation, ex));
+      } catch (IOException e) {
+        status.add(new Status(IStatus.ERROR, context.getBundle().getSymbolicName(), "Cannot extract bundle symbolicName or version " + bundleLocation, e));
+      }
+    }
+    status.addAll(startBundles(bundlesToStart.toArray(new Bundle[bundlesToStart.size()])));

```

Comment:
toArray? you can send the list as-is to startBundles, which is private, so no risks of doing nasty things with it

Reply:
Doh! fixed